### PR TITLE
[KIP-113] BLS public key registry

### LINKED ISSUE #113: Discussion for Simple BLS Registry
This issue is created for discussions on how to make interfaces for BLS registry.

BLS registry design is the first step to bring BLS crypto in Klaytn.
Further usages will be described in the following KIPs.

This issue number will be used as a KIP number.

### PR DESCRIPTION
**Pull request by ian0371**
KIP-113 is a proposal for BLS public key registry.
[See this PR in markdown](https://github.com/klaytn/kips/blob/kip113/KIPs/kip-113.md)

Links are to be added in implementation, test cases sections.
Closes #113

### CODE REVIEW

**Comment by ian0371** [comment]
I have read the CLA Document and I hereby sign the CLA

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]: Can you more elaborate what kinds of authentication and authorization will be done in this function?

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md:94]: If it is not too complicated, what about adding description in the solidity code as a comment?

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]: If we say that "a Klaytn node specify the target block number", this "target block number" is a configurable parameter per node? It seems like we need to define this value as a chain configuration, not a node configuration.

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
- kip113CompatibleBlock: the block number which a Klaytn node starts using Registry
```

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
- kip113ContractAddress: the address of a contract that implements IKIP_113.
```

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
4. chainConfig: deploy the contract from EOA, and set the address to chainConfig (the method used for TreasuryRebalance, a.k.a. [KIP-103](https://kips.klaytn.foundation/KIPs/kip-103))
```

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]: It seems like we decided to use chainConfig. Can you explain why? When I see this table, chainConfig is not the best option.

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: It's chain-specific so I wrote it in the below "cypress-specific implementation" section. It's not mandatory so service chain operators can decide to not perform authentication/authorization.

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: should i also update "Registry" in `kip113CompatibleBlock`?

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: No, "the target block number" indicates a predefined block number. I think it would have been "a target block number" if it was node-specific. BTW, [KIP-103](https://github.com/klaytn/kips/blob/main/KIPs/kip-103.md) also mentioned it like this.

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: fixed in 4fed377

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: fixed in 4fed377

File: KIPs/kip-113.md
Comment (baerae-colin) [reply KIPs/kip-113.md]: @ian0371 Then, what about replacing "a Klaytn node" with "a Klaytn chain"?
Or, we can also rephrase the sentence like:

The target block number must be specified per chain using chainConfig;

File: KIPs/kip-113.md
Comment (baerae-colin) [reply KIPs/kip-113.md]: Yeah, but it is not easy to be linked the things. What about adding some more descriptions like:

```suggestion
- The function SHOULD authenticate and authorize `msg.sender`. The implementation can be different from chain to chain. For Cypress (Klaytn's mainnet), please refer to the section `Cypress-specific implementation`.
```

File: KIPs/kip-113.md
Comment (baerae-colin) [reply KIPs/kip-113.md]: @ian0371 It would be also good if you agree with the change.

Comment (ian0371) [reply KIPs/kip-113.md]: let's leave it for brevity

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: Updated the table and added rationale in 5282741. PTAL

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: Added in 3f5f882

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: fixed in 1b90a54

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md:94]: added in 3033819

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]: Comparing to other implementations, IKIP113 is more commonly used until now rather than IKIP_113.

```suggestion
interface IKIP113 {
```


https://github.com/klaytn/klaytn-contracts/blob/master/contracts/KIP/interfaces/IKIP13.sol

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
In terms of explicit contract logic upgrade, ChainConfig, Precompiled-contract, and Bytecode-injection are recommended.
```

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
##### Usability from ServiceChain
```

Comment (baerae-colin) [reply KIPs/kip-113.md]: Previously we decided to use the therm `ServiceChain`.

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
In terms of usability from ServiceChain, HeaderGov, ContractGov, ChainConfig, and Bytecode-injection are recommended.
```

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
In terms of various remaining aspects, ChainConfig, Precompiled-contract, and Bytecode-injection are recommended.
```

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]: I don't get the point of this statement. Can you explain little bit more why Klaytn nodes are main users? Klaytn nodes vs smart contracts seems little bit weird. Are you considering protocol vs smart contract?

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
- ease of use in all of ServiceChain, Baobab (testnet), Cypress (mainnet)
```

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: I meant users as in Registry readers. Registry is mainly read by Klaytn nodes like GovParam or KIP103. I'll clarify the term "user"

Comment (ian0371) [reply KIPs/kip-113.md]: let's use the term "subscriber"

Comment (baerae-colin) [reply KIPs/kip-113.md]: @ian0371 What about using the term "Core protocol" instead of "Klaytn nodes"?

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
- `registerPublicKey` reverts if `msg.sender` is not a valid `cnNodeId` registered in AddressBook.
```

There could be some dummy cnNodeId in the address book. Is this okay?

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
- `unregisterPublicKey` reverts if `msg.sender` is not an admin of AddressBook.
```

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: Registry can't tell if it's a dummy or not, so it won't revert.

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: Fixed in 86a1721 along with other bullet points

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: fixed in 6853068

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
author: Ian (@ian0371), Ollie (@blukat29), Joseph (@jiseongnoh), and Aidan (@aidan-kwon)
```

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]: Will you merge this after this Github link is updated?

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
| Hardfork required when upgrade                                  | X             | X                                      | X                 | O               | O                        | O                      |
```

Comment (baerae-colin) [reply KIPs/kip-113.md]: @ian0371 I am not sure the meaning is what you intended, but the previous term is also misleading. Let me know the rephrased term is right or not.

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
| Possible to set the target block number for address change                     | O             | O                                      | X                 | O               | O                        | O                      |
```

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]: Awareness of subscriber contracts after address update
-> This means that the subscriber contracts will notify after the address update?

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]: hmm immediate panic is a good thing?

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
| Possible to setup initial data before hardfork        | O                             | O                             | O                             | O                    | O                        | X                      |
```

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
| Possible to use a special address (like 0x0…400)                     | X<br> vanity address          | X<br> vanity address          | X<br> vanity address          | X<br> vanity address | O                        | O                      |
```

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: yep that's correct

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: The question is: will the subscriber contracts be able to use the latest contract address?
e.g., for [3] spare contract, subscriber contracts can use `address registry = IAddressBook(0x400).spareContract()`

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: I think the sooner you find out you didn't upgrade the node, the better. If it panics later, there can be situations where you have to rollback to the hardfork (e.g., incompatible DB).

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: I'll update the link but it's a private repo for now

**Comment by ian0371** [comment]
To enable flexible naming of parameters in ChainConfig, parameters section is added and chain configuration part is updated.
Instead of specifying the actual name for the configuration, it simply mentions that the parameters should somehow be written in ChainConfig.
(like https://eips.ethereum.org/EIPS/eip-2929#parameters)

So we can name the hardfork block like `mantleCompatibleBlock`.

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
| Possible to find the updated addresses from the subscriber contracts     | X             | △<br> Latest GovParam address required | O                 | X               | O                        | X                      |
```

File: KIPs/kip-113.md
Comment (baerae-colin) [reply KIPs/kip-113.md]: I suggested the phrase in another comment.

File: KIPs/kip-113.md
Comment (baerae-colin) [reply KIPs/kip-113.md]: I got your point, but it would be good to rephrase it.
```suggestion
| Possible to alert immediately for out-of-date nodes | △<br> At GovernanceData block | X<br> At the activation block | △<br> At SpareContract update | O                    | O                        | O                      |
```

Comment (ian0371) [reply KIPs/kip-113.md]: you mean `immediately at hardfork` right?

Comment (ian0371) [reply KIPs/kip-113.md]: do you mean alert as in notify/warn?

Comment (ian0371) [reply KIPs/kip-113.md]: I think "alerting out-of-date nodes" is ambiguous; who alerts to whom and how?

Comment (baerae-colin) [reply KIPs/kip-113.md]: okay then, can you suggest something else? 

1. I want the statement to be more yes/no question.
2. what to emphasize is that whether we have a method to force them update the node. Otherwise, the node will not work. 

So, my another suggestion would be

"Possible to panic the node if out-dated"

Comment (ian0371) [reply KIPs/kip-113.md]: Initially the answers were
- O if panics at hardfork block,
- △ if panics later at a predictable block,
- X if panics at undetermined block.

If the question is "Possible to panic outdated nodes", they they're all yes. (they will panic some day)
So I'll (1) change the question to "Possible to panic outdated nodes at hardfork" and (2) change answers from △ to X. (see fbd69de)

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: fixed in fbd69de

File: KIPs/kip-113.md
Comment (aidan-kwon) [inline KIPs/kip-113.md]: I think it is time to consider a general way to introduce a system contract. I will prepare an initial discussion material for this topic and share it with core developers in another channel. 
If we conclude on a specific way for system contract introduction, then this chapter doesn't need to be included in this spec. If not, let's proceed with your suggestion.

Comment (aidan-kwon) [reply KIPs/kip-113.md]: I think it can take a few weeks. If the discussion takes long, merge this PR and then I will make another PR for the change

Comment (ian0371) [reply KIPs/kip-113.md]: I agree with having a standardized way for system contracts (i.e., in another KIP).
Later, we can remove this section and make readers refer to the new KIP.

**Review by aidan-kwon — APPROVED** [review]

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
This standard defines an interface for BLS12-381 registry, which enables nodes or smart contracts to access and view all the public keys of GC members. This is not to replace AddressBook, but rather to supplement it in order to facilitate the use of BLS in various components, including signature-based random (KIP-114).
```

GC is the most appropriate term that we can use? Is there any reason that we can also store non-GC member's public key?

Comment (ian0371) [reply KIPs/kip-113.md]: We can use validators, or consensus nodes instead of GC.
There's no reason to store non-GC member key; KIP-113 registry only allows CN nodes registered in AddressBook to register their public key.

Comment (baerae-colin) [reply KIPs/kip-113.md]: Then, please use the term GC members rather than GCs. Thanks~

File: KIPs/kip-113.md
Comment (baerae-colin) [reply KIPs/kip-113.md]: Sorry for the late reply. So, in this case, you think panicing at the hard fork block is the best one, dont't you?

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
In other words, if Registry should be upgraded, Registry must be redeployed to have a different address.
```

File: KIPs/kip-113.md
Comment (baerae-colin) [inline KIPs/kip-113.md]:
```suggestion
- `unregisterPublicKey(cnNodeId)` reverts if `msg.sender` is not an admin of AddressBook, or if `cnNodeId` is not in AddressBook.
```

**Comment by ian0371** [comment]
The initial design was to provide a temporary BLS key storage for Klaytn v1.11, and was to be superseded by permissionless contract afterward.
I'll redesign this interface so the candidate list contract, a new contract that will be released with Klaytn v1.12, conforms to KIP-113.

**Comment by baerae-colin** [comment]
@ian0371 Please check out the remaining comments.

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: yes, i do think.

I removed this section, and we can further discuss in the upcoming KIP.

**Comment by ian0371** [comment]
The purpose of this KIP has changed:
- before: supplement AddressBook so nodes can read public keys of GC members
- after: define a generic interface so users can associate an address with a public key using the interface

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: This section is removed because it no longer defines how the registry as in system contract will be deployed. Any developers can deploy them in an upgradeable way.

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: This section is removed because it no longer defines how the registry as in system contract will be deployed, and thus not limited to storing public keys of GC members.

File: KIPs/kip-113.md
Comment (ian0371) [reply KIPs/kip-113.md]: This section is removed because a representative registry will not be deployed in Cypress as a system contract.

**Review by baerae-colin — APPROVED** [review]

**Pull request merged**
